### ISSUE TITLE
(a11y) Keyboard users can't access all content in Steps

### ISSUE THREAD

**Issue by bbsmooth** [freeCodeCamp/freeCodeCamp#46259]
### Describe the Issue

There are two bugs related to this issue. Both assume that the user is only using the keyboard to navigate the page and interact with the content.

1. Keyboard users are not able to access the result/hint after triggering the Check Your Code button when the button is near the bottom of the page and there are no editor lines below the Check Your Code button.

2. Keyboard users are not able to read all of the instructions in a Step when the instructions take up a considerable amount of the vertical space of the view port. This would generally happen when the user has the page zoomed in.

### Affected Page

https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-css-grid-by-building-a-magazine/step-75

### Steps to Reproduce

Using the above link (Learn CSS Grid by Building a Magazine - Step 75) and only the keyboard:

Bug 1:
1. Make sure keyboard focus is in the editor.
2. Type `Ctrl + m` to enable Tabbing out of the editor.
3. Hit Tab once to set focus on the Check Your Code button. The button should scroll into view if it was not already.
4. Trigger the button.
5. Try to scroll down to read the hint. 

Bug 2:

NOTE: My monitor is 1920 x 1200 and so I am basing these instructions on that size or smaller. If you have a bigger monitor (especially height-wise) then you might not be able to replicate this bug.

6. Expand your browser to take up the entire screen.
7. Increase the page zoom to 400%. This should put you in the tabbed mobile layout.
8. Make sure the keyboard focus is in the editor (you can use your mouse to do this).
9. Try to read all of the content in the instructions. You will most likely find that at least one of the lines in the middle is not accessible because the jump between line 143 and 144 is too big.

### Expected behavior

[WCAG 2.1.1](https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html) requires that all content be accessible via the keyboard, which the first issue fails. [WCAG 1.4.10](https://www.w3.org/WAI/WCAG21/Understanding/reflow.html) requires that the user be able to zoom up to 400% (on a 1280 wide view port) and still access all of the content, which is not possible for keyboard only users unless they have extraordinarily large screens, so the second issue fails both 1.4.10 and 2.1.1.

### System

N/A

### Additional context

The main culprit here appears to be that scrolling is done by line number in the editor. Obviously, fixing the scrolling so that it takes into account the content that is not technically in the editor but is displayed within the editor would work but I'm not sure how feasible that is? Alternatively, to make the instructions more accessible, we could add the ability to pop out the instructions into their own tab in the tablist. Perhaps even make that a user setting so the user wouldn't have to do it manually every time? My first thought for the results/hint was to ensure there is always a line number below them so that the user will always be able to scroll below them, but at 400% zoom I think they would still be too tall and we'd run into the same line gap issue as with the instructions. Perhaps adding a "maximize" button for the results so that the user can pop them out to a modal? Or we do have the "Tests" tab, but it only shows the results, not the hints. Perhaps add the hints to that tab?

**Comment by bbsmooth** [comment]
The more I think about this, I don't think that zoom is actually necessary to demonstrate the second bug. With the default 100% zoom it is still possible to have the instructions be too long and skip on keyboard scroll if you reduce the height of the view port. Granted, this takes a tall instructions box and a short view port, but the point is that we will have a variety of users with different view port heights and different zoom settings and we should be able to accommodate all of them by allowing them to scroll all content one line at a time.

**Comment by bbsmooth** [comment]
I think PR #46272 would technically be enough to address Bug 1. While the user would not be able to use the arrow keys to scroll down to see the result/hint, they would still be able to bring them into view by Tabbing to the "Ask for Help" button below. But in order to rely on this solution at least one of the buttons would always need to be available below the results/hints.

**Comment by bbsmooth** [comment]
PR #46272 did not address Bug 1 sufficiently since the "Ask for Help" and "Reset Step" buttons do not appear until after a certain amount of failed checks.

**Comment by Srisuma13** [comment]
@naomi-lgbt @bbsmooth I would like to work on this if it is up for grabs.

**Comment by bbsmooth** [comment]
@Srisuma13 Well, I'm not sure an actual solution has been decided on this yet since I'm the only one who has commented on this. I'm kind of surprised that this has a help wanted tag on it. I threw out a few possibilities but no one else chimed in, so I don't know what would be an acceptable solution or I would probably have just implemented it myself a while ago. Have you looked into this yourself? Do you have some suggestions on how to solve these keyboard accessibility issues?

**Comment by ShaunSHamilton** [comment]
Would someone remind me: Do we still have this functionality? _I only tested the linked challenge_

> ... the "Ask for Help" and "Reset Step" buttons do not appear until after a certain amount of failed checks.

---

@bbsmooth Please confirm/deny: The best solution is for the editor jaws to be in the tab-index flow of the editor document?

Would this be too confusing, as it would prevent Campers from getting the full context of the code without also tabbing into the jaws? Specifically:

```html
<p>As a keyboard user / screen-readerer, I want to navigate through</p>
<p>all the HTML code. However, the code is</p>

<!-- LOWER JAW -->
<!-- BUTTON: Check your code -->
<!-- Here is a hint -->

<p>split by the lower jaw. So, this is not zen</p>
```

**Comment by bbsmooth** [comment]
@ShaunSHamilton Whoops, almost let your question slip by. Actually, I'm not sure I understand exactly what you are asking here? I'm not blaming you for that. Can you give me the Dummies version :slightly_smiling_face: I think what I'm not understanding is this statement:

> The best solution is for the editor jaws to be in the tab-index flow of the editor document

**Comment by bbsmooth** [comment]
> Would someone remind me: Do we still have this functionality?

It appears that the Reset button now appears from the very start, it is only the Ask for Help button that does not appear until a certain amount of code checks take place. But I am realizing now that this still won't fix Bug #1, especially if the user is zoomed in quite a bit. I will try to clarify these issues a little further and provide videos to demonstrate them.

**Comment by bbsmooth** [comment]
This is an example of Bug 2. 

![simplescreenrecorder-2023-01-07_12 21 57](https://user-images.githubusercontent.com/5926192/211174130-5a37f018-7f61-4233-b8f7-133337855f1c.gif)

This is on a 1900 x 1200 screen. The browser is zoomed in 260%. Notice how you cannot read the entire instructions because the jump from line 143 to 144 is too big. Granted, eventually the bottom row of buttons will be removed, and we could somehow allow the user to hide some of the fixed controls above to create more space for the instructions. But WCAG requires that the user should be able to zoom up to 400%, and at that level of zoom we will not be able to create enough vertical space for the instructions no matter what we remove.

My two suggestions for dealing with this issue:

- Offer a mechanism for the user to open the instructions is a modal dialog so they can scroll the instructions line by line.
- Allow the user to separate the instructions from the editot into a separate tab. Then we will have one tab for just the instructions and one tab just for the code editor. Both will be scrollable line by line. Probably make this a setting so that it persists.

**Comment by ShaunSHamilton** [comment]
> Actually, I'm not sure I understand exactly what you are asking here?

Basically, the text itself should be important enough to be tabbed into.

We can tab to the "Run Code" button, and then the next tab is the "Reset Code" button. Instead, I am wondering if we should set the hint text to be the next tabbable region?

---

Looking at it again, though, I suspect you are correct in mentioning the scrolling as being the issue; it might be more sensible to change the scrolling with arrowkeys behaviour such that the whole editor can be "smoothly" viewed.

Could we make it such that:
1) Editor in focus
2) `Ctrl + m`
3) `Arrowdown` controls scroll bar instead of cursor

**Comment by bbsmooth** [comment]
Thank you @ShaunSHamilton, I think I understand what you are suggesting now. I will play around with this and see if I can find an accessible way to do what you are suggesting.

**Comment by bbsmooth** [comment]
> it might be more sensible to change the scrolling with arrowkeys behaviour such that the whole editor can be "smoothly" viewed.

That's an interesting idea as well. I'm going to steal it as my third suggestion :slightly_smiling_face: 

Hopefully monaco allows us some way to alter the behavior of the scroll bar. I will look into it.

**Comment by bbsmooth** [comment]
I just discovered that if the keyboard focus is in the monaco editor then you can force it to scroll one line at a time through the instructions and results by using the <kdb>Ctrl</kbd> modifier key with the up/down arrow keys. I have tested this with Chrome, Firefox, and Edge on both Windows 11 and Linux. It might be a good idea to have someone test on a Mac and verify which modifier keys should be used. This built-in functionality does appear to fix both "bugs" mentioned in the original post. 

The only potential gotcha here is if the keyboard focus is on a button in the results box (e.g Check Your Code button) then using the arrow keys will have no effect, even if the <kbd>Ctrl</kbd> modifier is used. The keyboard focus must be on a line number in the editor in order to scroll with the keyboard.

**Comment by bbsmooth** [comment]
@ArielLeslie has confirmed that the equivalent key combination on a mac is Cmd + PgUp/PgDn.

So I think we can consider this issue closed as there is a way for keyboard users to scroll through the instructions and results one line at a time. I think there are other issues that need to be addressed that are related to this, such as how do we let keyboard users no that this feature exists in the first place. But those should be opened as separate issues.
